### DISCUSSION TITLE
Protocol for generic callback

### DISCUSSION THREAD

**Discussion by ronf** [python/typing#927] [Q&A]
I have a function which implements a generic "poll" capability, where a provided callback is called repeatedly until it returns a result, with a small sleep between attempts. I've been trying to define the generic callback as a Protocol, but I can't seem to make it work. Here's a simplified example, with the actual polling functionality removed for conciseness:

```python
from typing import Protocol, TypeVar

_Result = TypeVar('_Result', covariant=True)

class _Func(Protocol[_Result]):
    def __call__(self, *args: object) -> _Result: ...

def foo(a: int, b: str) -> bool:
    print(a, b)
    return True

def poll(func: _Func, *args: object) -> _Result:
    return func(*args)

print('Returned:', poll(foo, 1, 'a'))
```

When I run mypy on this, I get:
```
t.py:15: error: Argument 1 to "poll" has incompatible type "Callable[[int, str], bool]"; expected "_Func[Any]"
```

If I replace `_Func` with `Callable[..., _Result]`, it works, and that's actually sufficient for my use case, but there are other places I'd like to be able to use this approach on functions with args/kwargs, and so I'd like to figure out what I'm doing wrong here. Any thoughts?

**Comment by JelleZijlstra** [comment]
The problem here is that `_Func` accepts any *args, but the function you pass in only accepts two arguments.

The right way to handle this sort of thing is ParamSpec from https://www.python.org/dev/peps/pep-0612/, but unfortunately mypy doesn't support that yet.

In the meantime, one way that works for your simplified example would be to use `def poll(func: Callable[..., _Result]`.

**Reply by ronf** [reply]
Thanks. I had a feeling it was something like that. I think my other option might be to use @overload here, but I would have to explicitly list each of the possible signatures I want to support, and then have the implementation use *args to allow it to be shared. Listing the specific arguments rather than having mypy determine them automatically based on the function I pass in sort of defeats the point, though.
